PyTado/http.py

```diff
@@ -330,21 +338,76 @@ def _login(self) -> tuple[int, str]:
+            raise TadoException(e) from e
 
         if response.status_code != 200:
             raise TadoException(
-                f"Login failed for unknown reason with status code {response.status_code}"
+                f"Login failed. Status code: {response.status_code} and reason: {response.reason}"
             )
 
-        refresh_token = self._set_oauth_header(response.json())
-        id_ = self._get_id()
+        self._device_flow_data = response.json()
+        _LOGGER.debug("Device flow response: %s", self._device_flow_data)
+
+        user_code = urlencode({"user_code": self._device_flow_data["user_code"]})
+        visit_url = f"{self._device_flow_data['verification_uri']}?{user_code}"
+        self._user_code = self._device_flow_data["user_code"]
+        self._device_verification_url = visit_url
+
+        _LOGGER.info("Please visit the following URL: %s", visit_url)
+
+        expires_in_seconds = self._device_flow_data["expires_in"]
+        self._expires_at = datetime.now() + timedelta(seconds=expires_in_seconds)
+
+        _LOGGER.info(
+            "Waiting for user to authorize the device. Expires at %s",
+            self._expires_at.strftime("%Y-%m-%d %H:%M:%S"))
+        self._device_activation_status = "PENDING"
+
+    def device_activation(self) -> None:
+        """Activate the device and get the refresh token"""
+        if self._expires_at is None or self._device_flow_data is None:
+            raise TadoException("User has not yet authorized the device")
 
-        return id_, refresh_token
+        while True:
+            if datetime.timestamp(datetime.now()) > datetime.timestamp(self._expires_at):
+                raise TadoException("User took too long to enter key")
+
+            # Await the desired interval, before polling the API again
+            time.sleep(self._device_flow_data["interval"])
+            try:
+                token_response = self._session.request(
+                    method="post",
+                    url="https://login.tado.com/oauth2/token",
+                    params={
+                        "client_id": CLIENT_ID_DEVICE,
+                        "device_code": self._device_flow_data["device_code"],
+                        "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+                    }
+                )
+            except requests.exceptions.ConnectionError as e:
+                raise TadoException(e) from e
+
+            if token_response.status_code != 200:
+                # The user has not yet authorized the device, let's continue
+                if token_response.status_code == 400 and token_response.json()[
+                        "error"] == "authorization_pending":
+                    _LOGGER.info(
+                        "Authorization pending, waiting for user to authorize. Continue polling.")
+                    continue
+
+                raise TadoException(
```

Comment:
Should `_device_activation_status` be set to some error status in this instance?